File: ios/walletTests/Mocks/MockProviiWallet.swift
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust

import Foundation
@testable import ProviiWallet

#if canImport(ProviiSDK)
import ProviiSDK
#endif

// MARK: - MockProviiWallet

/// Test double for ProviiWalletProtocol. All methods return canned values by
/// default. Override individual closures to simulate error conditions or
/// inspect call arguments.
final class MockProviiWallet: ProviiWalletProtocol, @unchecked Sendable {

    // MARK: - Configuration

    var processQrChallengeResult: Result<String, Error> = .success("challenge-id-abc123")
    var createAgeProofResult: Result<String, Error> = .success("{\"proof\":\"valid\",\"challenge_id\":\"challenge-id-abc123\",\"credential_id\":\"cred-id-xyz789\"}")
    var submitProofResult: Result<Bool, Error> = .success(true)
    var listCredentialsResult: Result<[CredentialInfo], Error> = .success([])
    var finalizeAndStoreCredentialResult: Result<String, Error> = .success("stored-cred-id-001")
    var getCredentialResult: String? = "{\"id\":\"cred-id-xyz789\",\"credentialType\":\"primary\"}"
    var getDiagnosticInfoValue: DiagnosticInfo = DiagnosticInfo(
        sdkVersion: "0.0.0-test",
        appVersion: "1.0.0",
        platform: "iOS",
        proverInitialized: true,
        credentialCount: 0,
        storageAvailable: true,
        configEnvironment: "test",
        lastProofGenerated: nil
    )
    var processScannedQrResult: QrAction = .verificationChallenge(challengeJson: "{\"challenge_id\":\"challenge-id-abc123\"}")
    var processManualEntryResult: Result<String, Error> = .success("challenge-id-abc123")

    // MARK: - Call tracking

    var processQrChallengeCallCount = 0
    var createAgeProofCallCount = 0
    var submitProofCallCount = 0
    var finalizeAndStoreCredentialCallCount = 0

    // MARK: - ProviiWalletProtocol conformance

    func processQrChallenge(qrContent: String) throws -> String {
        processQrChallengeCallCount += 1
        switch processQrChallengeResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func processScannedQr(qrContent: String) throws -> QrAction {
        return processScannedQrResult
    }

    func processManualEntry(input: String) throws -> String {
        switch processManualEntryResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func createAgeProof(credentialId: String, challengeId: String) throws -> String {
        createAgeProofCallCount += 1
        switch createAgeProofResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func createAgeProofAuto(challengeId: String) throws -> String {
        switch createAgeProofResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func submitProof(proofJson: String) throws -> Bool {
        submitProofCallCount += 1
        switch submitProofResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func listCredentials() throws -> [CredentialInfo] {
        switch listCredentialsResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func getCredential(credentialId: String) throws -> String? {
        return getCredentialResult
    }

    func getDiagnosticInfo() -> DiagnosticInfo {
        getDiagnosticInfoValue
    }

    // swiftlint:disable:next function_parameter_count
    func finalizeAndStoreCredential(
        headerJson: String,
        dobDays: Int32,
        rBitsB64: String,
        label: String?,
        credentialType: String,
        nickname: String?
    ) throws -> String {
        finalizeAndStoreCredentialCallCount += 1
        switch finalizeAndStoreCredentialResult {
        case .success(let v): return v
        case .failure(let e): throw e
        }
    }

    func storeCredentialWithLabel(
        credentialJson: String,
        label: String?,
        credentialType: String,
        nickname: String?
    ) throws -> String {
        return "stored-cred-id-002"
    }

    func deleteCredential(credentialId: String) throws {}

    func deleteSandboxCredentials() throws {}

    func updateCredentialNickname(credentialId: String, nickname: String?) throws {}

    func storeCredential(credentialJson: String) throws -> String {
        return "stored-cred-id-003"
    }

    func importCredential(credentialJson: String) throws -> String {
        return "imported-cred-id-001"
    }

    func importCredentialWithType(credentialJson: String, credentialType: String, nickname: String?) throws -> String {
        return "imported-cred-id-002"
    }

    func initializeProver(pkBytes: Data) throws {}

    func isBiometricAvailable() -> Bool {
        return true
    }

    func setStorageHandle(handle: SecureStorageHandle) throws {}

    func setVerifierBaseUrl(baseUrl: String) throws {}

    func getVerifierBaseUrl() -> String {
        return "https://verifier.example.test"
    }

    func getConfig() -> WalletConfig {
        return WalletConfig(
            autoSelect: true,
            networkTimeout: 30,
            cacheProvingKeys: false,
            issuerApiUrl: "https://issuer.example.test",
            verifierApiUrl: "https://verifier.example.test",
            verifierApiKey: nil,
            verifierOrigin: nil,
            environment: "test",
            enableParallelProver: false,
            maxProverThreads: 1 as UInt8
        )
    }

    func updateConfig(config: WalletConfig) throws {}

    func getAvailableSlotCount() throws -> UInt8 {
        return 1
    }

    func hasValidCredential() -> Bool {
        return false
    }

    func hasCredentialSecrets(credentialId: String) throws -> Bool {
        return true
    }

    func getProvableCredentialsForChallenge(challengeId: String) throws -> [CredentialSuitability] {
        return []
    }

    func getVerificationStatus() -> VerificationStatus {
        return VerificationStatus.notStarted
    }

    func cancelVerification(challengeId: String) throws {}

    func cleanupExpiredChallenges() -> UInt32 {
        return 0
    }

    func cleanupExpiredCredentials() -> UInt32 {
        return 0
    }

    func calculateAgeFromDob(dobIso: String) throws -> UInt32 {
        return 25
    }

    func checkNetworkStatus() -> NetworkStatus {
        return NetworkStatus(connected: true)
    }

    func createProgressTracker() -> ProgressTracker {
        fatalError("createProgressTracker not implemented in MockProviiWallet")
    }

    func reportProgress(tracker: ProgressTracker, stage: ProgressStage, message: String) {}

    func debugPreflight(credentialId: String, challengeId: String) throws -> String {
        return "{\"status\":\"ok\"}"
    }

    func diagnoseProofFailure(credentialId: String, challengeId: String) throws -> String {
        return "{\"diagnosis\":\"none\"}"
    }

    func fetchChallengeByShortCode(shortCode: String) throws -> String {
        return "{\"challenge_id\":\"challenge-id-abc123\"}"
    }

    func fetchChallengeDetails(challengeId: String) throws -> String {
        return "{\"challenge_id\":\"\(challengeId)\"}"
    }

    func getChallengeDiagnostics(challengeId: String) throws -> String {
        return "{\"cutoff_days\":6570}"
    }

    func handleDeeplink(url: String) throws -> DeeplinkAction {
        return DeeplinkAction.scanChallenge(payloadJson: "{\"challenge_id\":\"challenge-id-abc123\"}")
    }

    func parseQr(qrContent: String) throws -> String {
        return "{\"type\":\"verification_challenge\"}"
    }

    func parseQrPayload(qrContent: String) throws -> String {
        return "{\"type\":\"verification_challenge\"}"
    }

    func validateQr(qrContent: String) throws -> Bool {
        return true
    }

    func emergencyZeroize() {}

    func refreshIssuerKeys(jwksJson: String) throws {}
}

File: android/app/src/test/java/app/provii/wallet/data/MockProviiWallet.kt
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust
package app.provii.wallet.data

import app.provii.wallet.sdk.CredentialInfo
import app.provii.wallet.sdk.CredentialSuitability
import app.provii.wallet.sdk.DeeplinkAction
import app.provii.wallet.sdk.DiagnosticInfo
import app.provii.wallet.sdk.NetworkStatus
import app.provii.wallet.sdk.ProviiWalletInterface
import app.provii.wallet.sdk.ProgressStage
import app.provii.wallet.sdk.ProgressTracker
import app.provii.wallet.sdk.QrAction
import app.provii.wallet.sdk.SecureStorageHandle
import app.provii.wallet.sdk.VerificationStatus
import app.provii.wallet.sdk.WalletConfig

/**
 * Test double for [ProviiWalletInterface]. Returns configurable canned values
 * so that WalletRepository flow tests exercise orchestration logic without
 * loading native FFI or the proving key.
 */
open class MockProviiWallet : ProviiWalletInterface {
    // MARK: - Configuration

    var processQrChallengeResult: Result<String> = Result.success("challenge-id-abc123")
    var createAgeProofResult: Result<String> =
        Result.success(
            """{"proof":"valid","challenge_id":"challenge-id-abc123","credential_id":"cred-id-xyz789"}""",
        )
    var submitProofResult: Result<Boolean> = Result.success(true)
    var listCredentialsResult: Result<List<CredentialInfo>> = Result.success(emptyList())
    var finalizeAndStoreCredentialResult: Result<String> = Result.success("stored-cred-id-001")
    var getCredentialResult: String? = """{"id":"cred-id-xyz789","credentialType":"primary"}"""
    private var _verifierBaseUrl: String = "https://verifier.example.test"

    private val diagnosticInfoValue =
        DiagnosticInfo(
            sdkVersion = "0.0.0-test",
            appVersion = "1.0.0",
            platform = "Android",
            proverInitialized = true,
            credentialCount = 0u,
            storageAvailable = true,
            configEnvironment = "test",
            lastProofGenerated = null,
        )

    // MARK: - Call tracking

    var processQrChallengeCallCount = 0
    var createAgeProofCallCount = 0
    var submitProofCallCount = 0
    var finalizeAndStoreCredentialCallCount = 0

    // MARK: - ProviiWalletInterface implementation

    override fun processQrChallenge(qrContent: String): String {
        processQrChallengeCallCount++
        return processQrChallengeResult.getOrThrow()
    }

    override fun processScannedQr(qrContent: String): QrAction {
        return QrAction.VerificationChallenge("""{"challenge_id":"challenge-id-abc123"}""")
    }

    override fun processManualEntry(input: String): String {
        return processQrChallengeResult.getOrThrow()
    }

    override fun createAgeProof(
        credentialId: String,
        challengeId: String,
    ): String {
        createAgeProofCallCount++
        return createAgeProofResult.getOrThrow()
    }

    override fun createAgeProofAuto(challengeId: String): String {
        return createAgeProofResult.getOrThrow()
    }

    override fun submitProof(proofJson: String): Boolean {
        submitProofCallCount++
        return submitProofResult.getOrThrow()
    }

    override fun listCredentials(): List<CredentialInfo> {
        return listCredentialsResult.getOrThrow()
    }

    override fun getCredential(credentialId: String): String? {
        return getCredentialResult
    }

    override fun getDiagnosticInfo(): DiagnosticInfo {
        return diagnosticInfoValue
    }

    override fun finalizeAndStoreCredential(
        headerJson: String,
        dobDays: Int,
        rBitsB64: String,
        label: String?,
        credentialType: String,
        nickname: String?,
    ): String {
        finalizeAndStoreCredentialCallCount++
        return finalizeAndStoreCredentialResult.getOrThrow()
    }

    override fun storeCredentialWithLabel(
        credentialJson: String,
        label: String?,
        credentialType: String,
        nickname: String?,
    ): String = "stored-cred-id-002"

    override fun storeCredential(credentialJson: String): String = "stored-cred-id-003"

    override fun importCredential(credentialJson: String): String = "imported-cred-id-001"

    override fun importCredentialWithType(
        credentialJson: String,
        credentialType: String,
        nickname: String?,
    ): String = "imported-cred-id-002"

    override fun deleteCredential(credentialId: String) {}

    override fun deleteSandboxCredentials() {}

    override fun updateCredentialNickname(
        credentialId: String,
        nickname: String?,
    ) {}

    override fun initializeProver(pkBytes: ByteArray) {}

    override fun isBiometricAvailable(): Boolean = true

    override fun emergencyZeroize() {}

    override fun refreshIssuerKeys(jwksJson: String) {}

    override fun setStorageHandle(handle: SecureStorageHandle) {}

    override fun setVerifierBaseUrl(baseUrl: String) {
        _verifierBaseUrl = baseUrl
    }

    override fun getVerifierBaseUrl(): String = _verifierBaseUrl

    override fun getConfig(): WalletConfig =
        WalletConfig(
            autoSelect = true,
            networkTimeout = 30uL,
            cacheProvingKeys = false,
            issuerApiUrl = "https://issuer.example.test",
            verifierApiUrl = "https://verifier.example.test",
            verifierApiKey = null,
            verifierOrigin = null,
            environment = "test",
            enableParallelProver = false,
            maxProverThreads = 1u,
        )

    override fun updateConfig(config: WalletConfig) {}

    override fun getAvailableSlotCount(): UByte = 1u

    override fun hasValidCredential(): Boolean = false

    override fun hasCredentialSecrets(credentialId: String): Boolean = true

    override fun getProvableCredentialsForChallenge(challengeId: String): List<CredentialSuitability> = emptyList()

    override fun getVerificationStatus(): VerificationStatus = VerificationStatus.NotStarted

    override fun cancelVerification(challengeId: String) {}

    override fun cleanupExpiredChallenges(): UInt = 0u

    override fun cleanupExpiredCredentials(): UInt = 0u

    override fun calculateAgeFromDob(dobIso: String): UInt = 25u

    override fun checkNetworkStatus(): NetworkStatus = NetworkStatus(connected = true)

    override fun createProgressTracker(): ProgressTracker {
        error("createProgressTracker not implemented in MockProviiWallet")
    }

    override fun reportProgress(
        tracker: ProgressTracker,
        stage: ProgressStage,
        message: String,
    ) {}

    override fun debugPreflight(
        credentialId: String,
        challengeId: String,
    ): String = """{"status":"ok"}"""

    override fun diagnoseProofFailure(
        credentialId: String,
        challengeId: String,
    ): String = """{"diagnosis":"none"}"""

    override fun fetchChallengeByShortCode(shortCode: String): String = """{"challenge_id":"challenge-id-abc123"}"""

    override fun fetchChallengeDetails(challengeId: String): String = """{"challenge_id":"$challengeId"}"""

    override fun getChallengeDiagnostics(challengeId: String): String = """{"cutoff_days":6570}"""

    override fun handleDeeplink(url: String): DeeplinkAction =
        DeeplinkAction.ScanChallenge("""{"challenge_id":"challenge-id-abc123"}""")

    override fun parseQr(qrContent: String): String = """{"type":"verification_challenge"}"""

    override fun parseQrPayload(qrContent: String): String = """{"type":"verification_challenge"}"""

    override fun validateQr(qrContent: String): Boolean = true
}
